homeassistant/components/xiaomi_aqara/binary_sensor.py

```diff
@@ -446,6 +457,14 @@ def parse_data(self, data, raw_data):
         )
         self._last_action = value
 
+        if value == "vibrate":
+            if self._unsub_set_no_vibration:
+                self._unsub_set_no_vibration()
+            self._unsub_set_no_vibration = async_call_later(
```

Comment:
Is this a stateless device?

Reply:
The HomeAssistant implemantation was stateless, this PR gives it a state.
In my opnion the device itself has a state, since it detects vibrations, therefore the state is vibrating=ON, or not vibrating=OFF.

You cann't directly read this state from the device, but it will continuesly push vibration messages each 1 minute while vibrating so in that way you know if it is continuesly vibrating or not. (The same way motion sensors are almost always implemented).

Reply:
Note that the device also has a battery level (%), amplitude of the vibration (um), tilt angle (degrees) and oriantation coordinates that are all represented by seperate entities under the same device. These properties can directly be read from the device

Reply:
Our rules are that we only represent the state coming from the device api. We don't invent state. So this should not be an enroty in Home Assistant but just custom events.

So we can't accept this PR as is.

Reply:
And additionally it also fires events when detecting a tilt, or free fall/drop.

Reply:
Sure you can achieve the same thing with an input_boolean and 2 automations (one to turn it on at the vibration event and one to turn it off when the first automation was triggered more than 2 minutes ago).

But the thing is that many (inexperianced) users that have a vibration sensor probably want to know if the sensor is vibrating or not, that is kind of the point of a vibration sensor.
Now all those users will need to figure out how to achieve this and have an aditional 3 entities (1x input_boolean, 2x automation, or be even smarter or write more complicated automation or something to do it in less entities)
I made this PR mostly for other people that start with HomeAssistant and don't want to figure out how to get a simple "is it vibrating?" state based on events. (and because I also don't like having an aditional 3 entities that will clutter my HomeAssistant).

I do get that we need to oblige by the standards/rules set out for HomeAssisitant in order to have a uniform experiance accros diffrent manufactures/devices. But I think this is beeing a bit to strict.

Reply:
Besides the vibration sensor itself is a battery powered zigbee device. Therefore to save battery it cannot sent states much faster than once a minute (what it does) to the gateway (this is also the case in many motion sensor).
Therefore if you want to know the "state" of a battery powered vibration sensor, you will need to have the logic "if no new vibration message within x minutes, turn state to off".

So basically if that logic is build in the gateway itself and HomeAssistant reads the state from the gateway it is fine.
But if HomeAssistant uses the same logic in its code, it is not.
I think that is pretty strict...

If this logic would be implemented in the upstream python library, would it then be considered okay for HomeAssistant?
I am willing to re-write the PR to shift the code to the upstream lib.

Reply:
Yes, it's ok to have the library do something like that.

Reply:
I looked into the PyXiaomiGateway library, that lib does not have any internall states or classes per device.
It just captures the data sent by the gateway and issues a callback to HomeAssistant.
HomeAssistant does all the data processing for all subdevices as it is now.

Therefore it would require a complete re-write of the upstream library.
That is out of scope for this PR.